Set memory usage level draft (#1533)
MicrosoftEdge/WebView2Feedback: specs/SetMemoryUsageLevel.md

We are introducing WebView2 API to reduce memory usage only for this type of inactive WebViews.

# Description
You may call the `SetMemoryUsageLevel` API to have the WebView2 consume less memory and going back to normal usage. This is useful when your Win32 app becomes invisible, 
but still wants to have script running or monitoring requests from network.

# Examples
## .NET, WinRT
```c#
async protected void OnBecomingInactive()
{
    webView.CoreWebView2.SetMemoryUsageLevel(CoreWebView2MemoryUsageLevel.Idle);
}
async protected void OnBecomingActive()
{
    webView.CoreWebView2.SetMemoryUsageLevel(CoreWebView2MemoryUsageLevel.Normal);
}
```
## Win32 C++
```cpp
bool ViewComponent::HandleWindowMessage(
    HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam, LRESULT* result)
{
    if (message == WM_SYSCOMMAND)
    {
        if (wParam == SC_MINIMIZE)
        {
            OnBecomingInactive();
        }
        else if (wParam == SC_RESTORE)
        {
            OnBecomingActive();
        }
    }
}

void ViewComponent::OnBecomingInactive()
{

>     HRESULT hr = webView->SetMemoryUsageLevel(COREWEBVIEW2_MEMORY_USAGE_LEVEL_IDLE);

Reviewer: Didn't we have a previous sample that did TrySuspend / Resume in this same sort of code? Does it make sense for this sample to show when to use SetMemoryUsageLevel versus TrySuspend?

Author: I would expect an specific app to use TrySuspend/Resume or SetMemoryUsageLevel, not both. Will add some pseudo code there to do suspend with a CanSuspend check

Author: Updated the sample.

    if (FAILED(hr))
        ShowFailure(hr, L"Failed to set SetMemoryUsageLevel to idle");
}

void ViewComponent::OnBecomingActive()
{
    HRESULT hr = webView->SetMemoryUsageLevel(COREWEBVIEW2_MEMORY_USAGE_LEVEL_NORMAL);
    if (FAILED(hr))
        ShowFailure(hr, L"Failed to set SetMemoryUsageLevel to normal");
}

```

# Remarks
See [API Details](#api-details) section below for more details.

# API Notes
See [API Details](#api-details) section below for API reference.

# API Details

## Win32 C++
```IDL
/// Specifies memory usage level of WebView.
[v1_enum]
typedef enum COREWEBVIEW2_MEMORY_USAGE_LEVEL {
    /// Specifies normal memory usage level.
    COREWEBVIEW2_MEMORY_USAGE_LEVEL_NORMAL,

>     /// Specifies idle memory usage level.

Reviewer: Naming this Idle feels a little weird to me. Idle is naming it after a scenario that can use this rather than naming it after what it does. It sort of seems misleading to me since we're not changing the memory usage to idle. Instead for an idle app scenario you want to change memory usage level to a lower value. What about Reduced or Low?

Or name it to say more what the tradeoff is, like PreferSwapDiskOverInMemory (is that what this does?)

There's an existing enum Windows.System.AppMemoryUsageLevel with values Low, Medium, High, OverLimit. Not totally applicable here.

Author: Will change to Low. This is for expressing intent, not used memory. Should we change the name of the API to something like SetMemoryConsumptionTarget or some other name?

Reviewer: Yes SetMemoryConsumptionTarget and Low sound good to me!

Author: Changed it to MemoryUsageTargetLevel property with Low and Normal values.

  /// comsumption level of WebView. Scripts will not be impacted and continue to run.
  /// This is useful for inactive apps that still want to run scripts and/or keep
  /// network connections alive and therefore could not call `TrySuspend` and `Resume`
  /// to reduce memory consumption.
  /// These apps can set memory usage level to `COREWEBVIEW2_MEMORY_USAGE_LEVEL_IDLE` when
  /// the app becomes inactive, and set back to `COREWEBVIEW2_MEMORY_USAGE_LEVEL_NORMAL` when
  /// the app becomes active.
  /// It is not neccesary to set CoreWebView2Controller's IsVisible property to false when calling the API.
  /// It is a best effort operation to change memory usage level, and the API will return before the operation completes.
  /// Setting the level to `COREWEBVIEW2_MEMORY_USAGE_LEVEL_IDLE` could potentially cause
  /// memory for some WebView browser processes to be swapped out to disk when needed. Therefore,
  /// it is important for the app to set the level back to `COREWEBVIEW2_MEMORY_USAGE_LEVEL_NORMAL`
  /// when the app becomes active again to have smoothy user experience.
  /// Setting memory usage level back to normal will not happen automatically.

>   HRESULT SetMemoryUsageLevel([in] COREWEBVIEW2_MEMORY_USAGE_LEVEL level);

Author: Added description.

# Background
Some WebView2 apps want to continue to run scripts while inactive and therefore cannot make usage of `TrySuspend` and `Resume` APIs to reduce resource consumption.
We are introducing WebView2 API to reduce memory usage only for this type of inactive WebViews.

# Description
You may set the `MemoryUsageTargetLevel` property to have the WebView2 consume less memory and going back to normal usage. This is useful when your Win32 app becomes invisible, 
but still wants to have script running or monitoring requests from network.

# Examples
## .NET, WinRT
```c#
async protected void OnBecomingInactive()
{

>     // CanSuspendWebPage() checks whether the current web contents in WebView can be suspended.

Reviewer: CanSuspendWebPage in comment and CanSuspendWebView in code. Also would be good to specify this is an app dependent logic
```suggestion
    // CanSuspendWebPage() uses app specific logic to check whether the current web contents in the WebView2 can be suspended.
```

>     if (CanSuspendWebView()) {

Reviewer: All the sample code should use Microsoft coding style with braces on their own lines.
```suggestion
    if (CanSuspendWebView())
    {
```

  
  /// An app may set `MemoryUsageTargetLevel` to indicate desired memory
  /// comsumption level of WebView. Scripts will not be impacted and continue to run.
  /// This is useful for inactive apps that still want to run scripts and/or keep
  /// network connections alive and therefore could not call `TrySuspend` and `Resume`
  /// to reduce memory consumption.
  /// These apps can set memory usage target level to `COREWEBVIEW2_MEMORY_USAGE_TARGET_LEVEL_LOW` when
  /// the app becomes inactive, and set back to `COREWEBVIEW2_MEMORY_USAGE_TARGET_LEVEL_NORMAL` when
  /// the app becomes active.
  /// It is not neccesary to set CoreWebView2Controller's IsVisible property to false when calling the API.
  /// It is a best effort operation to change memory usage level, and the API will return before the operation completes.
  /// Setting the level to `COREWEBVIEW2_MEMORY_USAGE_TARGET_LEVEL_LOW` could potentially cause
  /// memory for some WebView browser processes to be swapped out to disk when needed. Therefore,
  /// it is important for the app to set the level back to `COREWEBVIEW2_MEMORY_USAGE_TARGET_LEVEL_NORMAL`

>   /// when the app becomes active again to have smoothy user experience.

Reviewer: ```suggestion
  /// when the app becomes active again to have a smooth user experience.
```

  /// Setting memory usage level back to normal will not happen automatically.
  /// An app should choose to use either the combination of `TrySuspend` and `Resume` or the combination

>   /// of setting MemoryUsageTargetLevel to low and normal. It is advices to not mix of them.

Reviewer: ```suggestion
  /// of setting MemoryUsageTargetLevel to low and normal. It is not advisable to mix them.
```

>   /// Suspend and resume opertaion is a superset of setting MemoryUsageTargetLevel.

Reviewer: ```suggestion
  /// The TrySuspend and Resume methods will change the MemoryUsageTargetLevel.
```
